### PR TITLE
[mlir][tensor] Add support for tensor.pack static shapes inference.

### LINKED ISSUE #16317: Llama-7b Crashes during IREE Compilation
### What happened?

While compilation, I'm getting the following crash:
```
iree-compile: /home/azureuser/work/iree-vivek/third_party/llvm-project/mlir/lib/Dialect/Tensor/IR/TensorOps.cpp:843: static void mlir::tensor::EmptyOp::build(mlir:
:OpBuilder &, mlir::OperationState &, ArrayRef<int64_t>, mlir::Type, mlir::Attribute): Assertion `all_of(staticShape, [](int64_t sz) { return !ShapedType::isDynamic(sz); }) && "expected only static sizes"' failed.                                                                                                                 Please report issues to https://github.com/openxla/iree/issues and include the crash backtrace.                                                                    
  #0 0x00007fd3a8bc5add llvm::sys::PrintStackTrace(llvm::raw_ostream&, int) /home/azureuser/work/iree-vivek/third_party/llvm-project/llvm/lib/Support/Unix/Signals.inc:723:11                                                                                                                                                           #1 0x00007fd3a8bc5fcb PrintStackTraceSignalHandler(void*) /home/azureuser/work/iree-vivek/third_party/llvm-project/llvm/lib/Support/Unix/Signals.inc:798:1       
  #2 0x00007fd3a8bc3ff6 llvm::sys::RunSignalHandlers() /home/azureuser/work/iree-vivek/third_party/llvm-project/llvm/lib/Support/Signals.cpp:105:5                   #3 0x00007fd3a8bc67e5 SignalHandler(int) /home/azureuser/work/iree-vivek/third_party/llvm-project/llvm/lib/Support/Unix/Signals.inc:413:1                        
  #4 0x00007fd39ca42520 (/lib/x86_64-linux-gnu/libc.so.6+0x42520)                                                                                                    #5 0x00007fd39ca969fc pthread_kill (/lib/x86_64-linux-gnu/libc.so.6+0x969fc)                                                                                     
  #6 0x00007fd39ca42476 gsignal (/lib/x86_64-linux-gnu/libc.so.6+0x42476)                                                                                          
  #7 0x00007fd39ca287f3 abort (/lib/x86_64-linux-gnu/libc.so.6+0x287f3)                                                                                            
  #8 0x00007fd39ca2871b (/lib/x86_64-linux-gnu/libc.so.6+0x2871b)                                                                                                  
  #9 0x00007fd39ca39e96 (/lib/x86_64-linux-gnu/libc.so.6+0x39e96)                                                                                                  
 #10 0x00007fd3b079dcf8 mlir::tensor::EmptyOp::build(mlir::OpBuilder&, mlir::OperationState&, llvm::ArrayRef<long>, mlir::Type, mlir::Attribute) /home/azureuser/wo
rk/iree-vivek/third_party/llvm-project/mlir/lib/Dialect/Tensor/IR/TensorOps.cpp:844:9                                                                              
 #11 0x00007fd3a98d1731 mlir::tensor::EmptyOp mlir::OpBuilder::create<mlir::tensor::EmptyOp, llvm::SmallVector<long, 6u>&, mlir::Type&>(mlir::Location, llvm::Small
Vector<long, 6u>&, mlir::Type&) /home/azureuser/work/iree-vivek/third_party/llvm-project/mlir/include/mlir/IR/Builders.h:500:5                                     
 #12 0x00007fd3ae30d36e mlir::linalg::GeneralizeOuterUnitDimsPackOpPattern::matchAndRewrite(mlir::tensor::PackOp, mlir::PatternRewriter&) const /home/azureuser/wor
k/iree-vivek/third_party/llvm-project/mlir/lib/Dialect/Linalg/Transforms/Transforms.cpp:1214:26                                                                    
 #13 0x00007fd3ac06246b mlir::detail::OpOrInterfaceRewritePatternBase<mlir::tensor::PackOp>::matchAndRewrite(mlir::Operation*, mlir::PatternRewriter&) const /home/
azureuser/work/iree-vivek/third_party/llvm-project/mlir/include/mlir/IR/PatternMatch.h:330:12                                                                      
 #14 0x00007fd3afec5886 mlir::PatternApplicator::matchAndRewrite(mlir::Operation*, mlir::PatternRewriter&, llvm::function_ref<bool (mlir::Pattern const&)>, llvm::f
unction_ref<void (mlir::Pattern const&)>, llvm::function_ref<mlir::LogicalResult (mlir::Pattern const&)>)::$_1::operator()() const /home/azureuser/work/iree-vivek/
third_party/llvm-project/mlir/lib/Rewrite/PatternApplicator.cpp:208:31
```

### Steps to reproduce your issue

IR is available at: https://gist.github.com/vivekkhandelwal1/185ed5eb3f76c7d88fd37a489185d0fb

Command to reproduce the crash: 
```
./build/tools/iree-compile --iree-hal-target-backends=llvm-cpu --iree-input-type=tm_tensor --iree-util-zero-fill-elided-attrs llama_7b_linalg_elided.mlir -o llama_7b_fp32.vmfb
```

### What component(s) does this issue relate to?

MLIR, Compiler

### Version information

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by hanhanW**
Fixes https://github.com/openxla/iree/issues/16317

### CODE REVIEW

File: mlir/lib/Dialect/Tensor/IR/TensorOps.cpp [hunk 1]
```diff
@@ -3983,6 +3983,41 @@ static bool paddingIsNotNeeded(PackOp op) {
                                       op.getMixedTiles());
 }
 
+// Returns true if the `srcShape` or `destShape` is different from the one in
+// `packOp`.
```
Comment (qedawkins) [inline mlir/lib/Dialect/Tensor/IR/TensorOps.cpp:3987]:
```suggestion
// Returns true if the `srcShape` or `destShape` is different from the one in
// `packOp` and populates each with the inferred static shape.
```

File: mlir/lib/Dialect/Tensor/IR/TensorOps.cpp [hunk 2]
```diff
@@ -4003,6 +4038,31 @@ LogicalResult PackOp::canonicalize(PackOp packOp, PatternRewriter &rewriter) {
...
+    if (destShape != packOp.getDestType().getShape()) {
+      auto newDestType = packOp.getDestType().clone(destShape);
+      dest =
+          rewriter.create<tensor::CastOp>(loc, newDestType, packOp.getDest());
+    }
+    Value newOp = rewriter.create<tensor::PackOp>(
+        loc, source, dest, packOp.getInnerDimsPos(), packOp.getMixedTiles(),
+        packOp.getPaddingValue(), packOp.getOuterDimsPerm());
+    rewriter.replaceOpWithNewOp<tensor::CastOp>(
+        packOp, packOp.getResult().getType(), newOp);
```
Comment (qedawkins) [inline mlir/lib/Dialect/Tensor/IR/TensorOps.cpp:4062]: Is this well suited for a canonicalization? I'm wondering about cases where a `pack` and `unpack` could have folded away but this pattern introduces a `tensor.cast` in the middle. Maybe we need the same pattern for unpack too?

Comment (hanhanW) [reply mlir/lib/Dialect/Tensor/IR/TensorOps.cpp:4062]: This is common in [Linalg](https://github.com/llvm/llvm-project/blob/d57515bd107bc76df5a042ffee2b7dc6125ffef1/mlir/lib/Dialect/Linalg/IR/LinalgOps.cpp#L2259-L2262) and I think we can add such functionality to tensor ops. Good point on pack/unpack folding. I think they can be folded if the order of applying patterns is correct. To make the result IR deterministic, we might need it for unpack. So yes, I will prepare a patch and send it out for review. Do you think it is better to land both patterns together? If so, I will put the update to the PR. If it does not matter, I will land it as a follow-up.

Comment (qedawkins) [reply mlir/lib/Dialect/Tensor/IR/TensorOps.cpp:4062]: If there is a possibility of non-deterministic IR probably best to land at the same time. In this case the pack-unpack canonicalization seems to always apply before this casting pattern so maybe it is ok here? Hard to predict exactly what the pattern applicator could do for any input IR though.

**Review by qedawkins — APPROVED** [review]
LGTM, contingent on addressing the discussed pack-unpack case in a follow up.

**Pull request merged**